velox/common/memory/Memory.h

```diff
@@ -65,89 +65,174 @@ constexpr uint16_t kDefaultAlignment = 64;
+
+  /// Tracks memory usage from leaf nodes to root and updates immediately
+  /// with atomic operations.
+  /// Unlike pool's getCurrentBytes(), getMaxBytes(), trace's API's returns
+  /// the aggregated usage of subtree. See 'ScopedChildUsageTest' for
+  /// difference in their behavior.
   virtual void setMemoryUsageTracker(
       const std::shared_ptr<MemoryUsageTracker>& tracker) = 0;
+
+  /// Returns the memory usage tracker associated with this memory pool.
   virtual const std::shared_ptr<MemoryUsageTracker>& getMemoryUsageTracker()
       const = 0;
-  // Used for external aggregation.
+
+  /// Used for external aggregation.
   virtual void setSubtreeMemoryUsage(int64_t size) = 0;
+
   virtual int64_t updateSubtreeMemoryUsage(int64_t size) = 0;
-  // Get the cap for the memory node and its subtree.
+
+  /// Used to manage existing externally allocated memories without doing a new
+  /// allocation.
+  virtual void reserve(int64_t /* bytes */) {
+    VELOX_NYI("reserve() needs to be implemented in derived memory pool.");
+  }
+
+  /// Sometimes in memory governance we want to mock an update for quota
+  /// accounting purposes and different implementations can
+  /// choose to accommodate this differently.
+  virtual void release(int64_t /* bytes */, bool /* mock */ = false) {
+    VELOX_NYI("release() needs to be implemented in derived memory pool.");
+  }
+
+  /// Get the cap for the memory node and its subtree.
   virtual int64_t getCap() const = 0;
-  virtual uint16_t getAlignment() const = 0;
 
-  ////////////////////// resource governing methods ////////////////////
-  // Called by MemoryManager and MemoryPool upon memory usage updates and
-  // propagates down the subtree recursively.
+  /// Called by MemoryManager and MemoryPool upon memory usage updates and
+  /// propagates down the subtree recursively.
   virtual void capMemoryAllocation() = 0;
-  // Called by MemoryManager and propagates down recursively if applicable.
+  /// Called by MemoryManager and propagates down recursively if applicable.
   virtual void uncapMemoryAllocation() = 0;
-  // We might need to freeze memory allocating operations under severe global
-  // memory pressure.
+  /// We might need to freeze memory allocating operations under severe global
+  /// memory pressure.
   virtual bool isMemoryCapped() const = 0;
 
-  ////////////////////////// tree methods ////////////////////////////
-  // TODO: might be able to move implementation to abstract class
-  // name of the node, preferably unique among siblings, but not enforced.
-  virtual const std::string& getName() const = 0;
-  virtual std::weak_ptr<MemoryPool> getWeakPtr() = 0;
-  virtual MemoryPool& getChildByName(const std::string& name) = 0;
-  virtual uint64_t getChildCount() const = 0;
-  virtual void visitChildren(
-      std::function<void(MemoryPool* FOLLY_NONNULL)> visitor) const = 0;
-
+ protected:
+  /// Invoked by addChild() to create a child memory pool object. 'parent' is
```

Comment:
Why do we need this method?

Reply:
This is the actual memory pool object creation and it is used by addChild() which is mostly independent of the actual pool implementation.